### PR TITLE
 Fully rewrite our sample app to be JWT & multi-region aware

### PR DESCRIPTION
**Pull request by jayrave**
https://fieldwire.atlassian.net/browse/REG-2078
- Deleted out existing flows for talking to the FW API
- Added more modular clients to be able to better organize & showcase what's possible

Note: please go through the commits one by one for easier reviewing

### CODE REVIEW

File: app/src/net/fieldwire/SampleCalls.java [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+            @NotNull String refreshToken,
+            @NotNull String accessToken
+    ) {
+        // Build gson to suit the needs of FW data deserialization
+        gson = new GsonBuilder()
+                .setDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSX")
+                .setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES)
+                .create();
+
+        // Build the components that takes care of managing our tokens
```
Comment (pirlande-fw) [inline app/src/net/fieldwire/SampleCalls.java]: typo: `Build the component...`

File: app/src/net/fieldwire/SampleCalls.java [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+        final String regionalBaseUrl;
+        if (region == Region.EU) {
+            regionalBaseUrl = RegionalEndpoint.API_BASE_URL_EU;
+        } else if (region == Region.US) {
+            regionalBaseUrl = RegionalEndpoint.API_BASE_URL_US;
+        } else {
+            throw new RuntimeException("Unrecognizable region: " + region);
+        }
+
+        // Build the impl that allows super endpoints to be called
```
Comment (pirlande-fw) [inline app/src/net/fieldwire/SampleCalls.java]: `that allows super endpoints` -> `that allows regional endpoints`

File: app/src/net/fieldwire/TokenManager.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        Response<JwtResponse> response = new Retrofit.Builder()
+                .baseUrl(SuperEndpoint.API_BASE_URL)
+                .addConverterFactory(GsonConverterFactory.create(gson))
+                .build()
+                .create(SuperEndpoint.class)
+                .createAccessToken(new CreateJwt(this.refreshToken))
+                .execute();
+
+        // NOTE: please make sure to persist this somewhere outside the script
+        // so it can reused for the next invocation of the script (otherwise,
```
Comment (pirlande-fw) [inline app/src/net/fieldwire/TokenManager.java]: nit: `so it can reused` -> `so it can be reused`

File: app/src/net/fieldwire/SampleCalls.java [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+
+        ProjectUsersResponse projectUsersResponse = tokenManager.executeWithToken(
+                authz -> endpoint.inviteUser(
+                    authz,
+                    project.id,
+                    new InviteUser(emailForProjectInvite)
+                )
+        );
+
+        ProjectUser user = projectUsersResponse.users.get(0);
```
Comment (pirlande-fw) [inline app/src/net/fieldwire/SampleCalls.java:174]: nit: could be great handling the case where there is no users in the response (I naturally put the email address of my account owner, which is ok but returns an empty array).

What about:
```
 if (projectUsersResponse.users.stream().count() == 0) {
      throw new RuntimeException("User already exists in the project, use a different email address " + emailForProjectInvite);
 }
```

File: app/src/net/fieldwire/models/response/AccountUser.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    public String toString() {
+        return "User{" +
+                "firstName='" + firstName + '\'' +
+                ", lastName='" + lastName + '\'' +
+                ", accountRole='" + accountRole + '\'' +
+                ", email='" + email + '\'' +
+                ", company='" + company + '\'' +
+                ", phoneNumber='" + phoneNumber + '\'' +
+                ", photoUrl='" + photoUrl + '\'' +
+                ", currentSignInAt=" + currentSignInAt +
```
Comment (pirlande-fw) [inline app/src/net/fieldwire/models/response/AccountUser.java:26]: nit: add single quotes to be consistent with other timestamps

File: app/src/net/fieldwire/models/response/AccountUser.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    public String email;
+    public String company;
+    public String phoneNumber;
+    public String photoUrl;
+    public Date currentSignInAt;
+    public int invitedById;
+
+    @Override
+    public String toString() {
+        return "User{" +
```
Comment (pirlande-fw) [inline app/src/net/fieldwire/models/response/AccountUser.java]: If I've read correctly, we never log `AccountUser` and `ProjectUser` directly from the sample code, but I guess this is for convenience if our customers want to debug these entities, right?

File: app/src/net/fieldwire/models/response/Project.java [hunk 7]
```diff
@@ -26,8 +14,15 @@ public String toString() {
...
+                ", updatedAt='" + updatedAt + '\'' +
+                ", deviceCreatedAt='" + deviceCreatedAt + '\'' +
+                ", deviceUpdatedAt='" + deviceUpdatedAt + '\'' +
+                ", deletedAt='" + deletedAt + '\'' +
+                ", createdAt='" + createdAt + '\'' +
+                ", updatedAt='" + updatedAt + '\'' +
+                ", id=" + id +
+                '}';
     }
 }
```
Comment (pirlande-fw) [inline app/src/net/fieldwire/models/response/Project.java]: nit: empty EOF

**Review by pirlande-fw — APPROVED** [review]
Super clear instructions, I've run the code easily on staging and tested both the US & EU region.

File: app/src/net/fieldwire/models/response/AccountUser.java (hunk 6 above)
Comment (jayrave) [reply app/src/net/fieldwire/models/response/AccountUser.java]: - True for `AccountUser` => added it just for the sake of completeness
- `ProjectUser` => this one gets logged as part of the regional flow highlighted below:

```java
ProjectUser user = projectUsersResponse.users.get(0);
log_info("Invited user to project", user.toString());
```

File: app/src/net/fieldwire/models/response/AccountUser.java (hunk 5 above)
Comment (jayrave) [reply app/src/net/fieldwire/models/response/AccountUser.java:26]: Thanks for pointing this out. Looks like the other timestamps were typed as `String` which lead to the quoting. I updated them instead to be `Date` & regenerated the `toString` method for the required classes

Addressed - https://github.com/Fieldwire/fieldwire_java_sample/pull/3/commits/04d3ffe0a859bc0689658e9da6f919904c78baa4

File: app/src/net/fieldwire/SampleCalls.java (hunk 4 above)
Comment (jayrave) [reply app/src/net/fieldwire/SampleCalls.java:174]: Good point! Addressed - https://github.com/Fieldwire/fieldwire_java_sample/pull/3/commits/6797ff1135c65d7cd21080eeb2fc18c73ac14247

**Pull request merged**